### PR TITLE
Absolute color swapping for Ozaria and Thang Editor extensions

### PR DESCRIPTION
**Pull request by AndrewJakubowicz**
## Features

Color customization will work differently for Ozaria than in CodeCombat. This PR provides more tooling for importing our tints and selecting colors.

 - Color Tab in the thang editor now includes a tiny stateless **hex to hsl** calculator.
      - This is included to ensure it is quick to calculate the hsl for tints from a given hex.

Screen Shot 2019-07-10 at 11 32 12 AM

 - Selecting a color in the grid of colors is extremely tedious. Now there is an input that lets you select the colors with a given hex. Just type in the hex and click select.

Screen Shot 2019-07-10 at 11 31 17 AM

 - Ozaria type now changes how color is applied to the thang. The color is directly applied to give us fine grained control over color customization. In legacy types the color is shifted but it is not precise enough for what we are doing now.

**Example of ozaria absolute color applied**
Screen Shot 2019-07-10 at 11 33 13 AM

**Example of legacy color behaviour (notice color is not red but original colors are shifted and averaged. Creates blue hair artifact and yellow armor although settings are the same as above image)**

Screen Shot 2019-07-10 at 11 34 29 AM

## How to test

**Be using a local development environment.**

1. Access [localhost:3000/editor/thang/captain](http://localhost:3000/editor/thang/captain)

2. Click on the colors tab.

3. Click on the team grouping and slide the hue slider. You should see the armor changing colors.

4. Click back on Main tab.

5. Add the property `ozaria: true` to the Captain thangtype.

6. Click back on the Colors tab. and repeat step 3. This time while sliding the hue slider the armor will be a solid color.

### CODE REVIEW

File: app/lib/sprites/SpriteBuilder.coffee [hunk 1]
```diff
@@ -143,7 +143,19 @@ module.exports = class SpriteBuilder
```
Comment (AndrewJakubowicz) [inline app/lib/sprites/SpriteBuilder.coffee]: This file is the only change that could affect end users. But the change will only affect ThangTypes with ozaria set to true.

File: app/views/editor/thang/hslCalculator.vue [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<script>
```
Comment (AndrewJakubowicz) [inline app/views/editor/thang/hslCalculator.vue]: This is a simple color calculator to help translate the hex colors to the hsl that we use.

File: app/views/editor/thang/ThangTypeColorsTabView.coffee [hunk 3]
```diff
@@ -62,6 +63,9 @@ module.exports = class ThangTypeColorsTabView extends CocoView
       $("#shape-buttons").hide()
       $("#saved-color-tabs").show()
 
+    # Attach a stateless calculator widget
+    new ColorCalculator({ el: '#color-calculator' })
```
Comment (jmif) [inline app/views/editor/thang/ThangTypeColorsTabView.coffee]: Is this cleaned up when we navigate away from the page?  We may want to register a reference to it and then call `destroy()` when the view is being unloaded.  I don't know enough about Vue to know if it'll be automatically garbage collected 🤷‍♂

File: app/lib/sprites/SpriteBuilder.coffee [hunk 4]
```diff
@@ -143,7 +143,19 @@ module.exports = class SpriteBuilder
...
+        @buildColorMapForGroup(colorGroups[group], config)
+
+  # Simpler Ozaria color mapper.
+  # Instead of color shifting we apply the color directly.
+  buildOzariaColorMapForGroup: (shapes, config) ->
+    return unless shapes.length
+    for shapeKey in shapes
+      shape = @shapeStore[shapeKey]
+      continue if not shape.fc?
+      @colorMap[shapeKey] = hslToHex([config.hue, config.saturation, config.lightness])
```
Comment (jmif) [inline app/lib/sprites/SpriteBuilder.coffee]: I'm not quite sure what color map is / what this is doing.  Code looks generally sound but can't comment on its functionality.  Maybe a shortish comment here describing the intention would provide some context?  Or a reference to other color map code that could provide it?

Comment (AndrewJakubowicz) [reply app/lib/sprites/SpriteBuilder.coffee]: Fair point. The colorMap key is a shape, and the value is the new color to use.

This method creates a large object which can be queried when the shapes are being drawn to see if an alternative color should be used.

Will add comment.

File: app/views/editor/thang/ThangTypeColorsTabView.coffee (hunk 3 above)
Comment (AndrewJakubowicz) [reply app/views/editor/thang/ThangTypeColorsTabView.coffee]: Will add a destroy call. Great pickup.

**Pull request merged**